### PR TITLE
Loop alignment: Fix loop size calculation to exclude IG's size that are marked as not aligned

### LINKED ISSUE #68510: Assertion failed 'paddingToAdd == paddingNeeded' in 'System.SpanTests.IndexOfAnyExceptTests`1[Byte][System.Byte]:SomeElementsDontMatch_ReturnsOffset(int,System.Int32[]):this'
pipeline: runtime-coreclr libraries-jitstress
job: net7.0-windows-Release-x86-CoreCLR_checked-jitstress1-Windows.10.Amd64.Open
test: System.Memory.Tests

https://dev.azure.com/dnceng/public/_build/results?buildId=1736001&view=ms.vss-test-web.build-test-results-tab&runId=46986296&paneView=debug

```
C:\h\w\A5DE098F\w\B5E40A10\e>set COMPlus 
COMPlus_JitStress=1
COMPlus_TieredCompilation=0

C:\h\w\A5DE098F\w\B5E40A10\e>call RunTests.cmd --runtime-path C:\h\w\A5DE098F\p 
----- start Mon 04/25/2022  8:03:28.98 ===============  To repro directly: ===================================================== 
pushd C:\h\w\A5DE098F\w\B5E40A10\e\
"C:\h\w\A5DE098F\p\dotnet.exe" exec --runtimeconfig System.Memory.Tests.runtimeconfig.json --depsfile System.Memory.Tests.deps.json xunit.console.dll System.Memory.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing 
popd
===========================================================================================================

C:\h\w\A5DE098F\w\B5E40A10\e>"C:\h\w\A5DE098F\p\dotnet.exe" exec --runtimeconfig System.Memory.Tests.runtimeconfig.json --depsfile System.Memory.Tests.deps.json xunit.console.dll System.Memory.Tests.dll -xml testResults.xml -nologo -nocolor -notrait category=IgnoreForCI -notrait category=OuterLoop -notrait category=failing  
  Discovering: System.Memory.Tests (method display = ClassAndMethod, method display options = None)
  Discovered:  System.Memory.Tests (found 2406 of 2425 test cases)
  Starting:    System.Memory.Tests (parallel test collections = on, max threads = 4)

Assert failure(PID 5760 [0x00001680], Thread: 4896 [0x1320]): Assertion failed 'paddingToAdd == paddingNeeded' in 'System.SpanTests.IndexOfAnyExceptTests`1[Byte][System.Byte]:SomeElementsDontMatch_ReturnsOffset(int,System.Int32[]):this' during 'Emit code' (IL size 817; hash 0x8f2948ff; FullOpts)

    File: D:\a\_work\1\s\src\coreclr\jit\emitxarch.cpp Line: 10144
    Image: C:\h\w\A5DE098F\p\dotnet.exe

----- end Mon 04/25/2022  8:04:21.14 ----- exit code -1073740286 ----------------------------------------------------------
```

@kunalspathak @dotnet/jit-contrib

### PR DESCRIPTION
**Pull request by kunalspathak**
As explained in https://github.com/dotnet/runtime/issues/68510#issuecomment-1117623799, there is a scenario when setting the loop back edge that we know we cannot align a loop and removes the `IGF_LOOP_ALIGN` flag without fixing the `igSize`. That affects the `loopSize` calculation depending on if it was done during loop alignment adjustment or during output. 

Edit:
> I have added a flag `IGF_REMOVED_ALIGN` but commented out the usage for now. I have added an assert to see if this was 
silently missed (because we would not align a loop for some heuristic reason). I will enable the usage of `IGF_REMOVED_ALIGN` after that.

There were test failures https://dev.azure.com/dnceng/public/_build/results?buildId=1752445&view=results so definitely we 
were hitting this scenario, but it was not failing because the loop might be big enough to not align. I have figured out another 
way in 3704be5c435edfadaeba69b9d3c5f302e8f2e118 to track the information of IGs whose align flag was removed instead of 
adding `IGF_REMOVED_ALIGN`.

#65690: In this case, we had 2 backedges of a loop that was marked for aligned. When we set the 1st backedge, we realize that it is enclosing a loop that is marked for alignment, so we unset the flag of current loop. We visit 2nd backedge and again try to unset the align flag and expect that it was not already unset. I have updated the assert using `removedAlignFlag` to say that either the align flag should be present, or it was present at one point and then we removed it.

#65988: We decide to not place `align` instruction in prolog, but we didn't remove the align flag which later causes problem. 2322a67031eaf62f1f832e24682e92935bbca256 fixes that.

Fixes: https://github.com/dotnet/runtime/issues/68510, https://github.com/dotnet/runtime/issues/65690 and https://github.com/dotnet/runtime/issues/65988

### CODE REVIEW

File: src/coreclr/jit/emit.cpp [hunk 1]
```diff
@@ -5179,16 +5183,35 @@ unsigned emitter::getLoopSize(insGroup* igLoopHeader, unsigned maxLoopSize DEBUG
...
             {
+                JITDUMP(" but ends with align instruction, taking off %u bytes.",
+                        emitComp->opts.compJitAlignPaddingLimit);
                 // The current loop size should exclude the align instruction size reserved for next loop.
                 loopSize -= emitComp->opts.compJitAlignPaddingLimit;
             }
         }
+        else if (igInLoop->igFlags & IGF_REMOVED_ALIGN)
+        {
+            assert("!Failed to remove 15 bytes");
```
Comment (Wraith2) [inline src/coreclr/jit/emit.cpp:5194]: Is this really always 15 bytes?

Comment (kunalspathak) [reply src/coreclr/jit/emit.cpp:5194]: It is `compJitAlignPaddingLimit` but this code is just for testing purpose. I will remove it after I see the CI result.

Comment (Wraith2) [reply src/coreclr/jit/emit.cpp:5194]: OK.
I'm interested to see if you can manage to remove instructions cleanly. I've been trying to do that with unneeded jmps but I've found that doing it breaks exception handling completely.

Comment (jakobbotsch) [reply src/coreclr/jit/emit.cpp:5194]:
```suggestion
            assert(!"Failed to remove 15 bytes");
```

Comment (kunalspathak) [reply src/coreclr/jit/emit.cpp:5194]: good catch. Thanks! :)

**Comment by kunalspathak** [comment]
So, we have several tests that hits the assert I added for testing which means that even today, we mismatch the actual vs. expected padding because of resolution blocks being added in between the loop making `bbNatLoopNum` unreliable.

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib , @BruceForstall

File: src/coreclr/jit/compiler.cpp [hunk 2]
```diff
@@ -5154,6 +5154,8 @@ void Compiler::placeLoopAlignInstructions()
     {
         // Adding align instruction in prolog is not supported
         // hence just remove that loop from our list.
+        fgFirstBB->bbFlags &= ~BBF_LOOP_ALIGN;
```
Comment (BruceForstall) [inline src/coreclr/jit/compiler.cpp:5157]: This is a case where you don't call `unmarkLoopAlign`?

File: src/coreclr/jit/emit.cpp [hunk 3]
```diff
@@ -5096,10 +5143,14 @@ unsigned emitter::getLoopSize(insGroup* igLoopHeader, unsigned maxLoopSize DEBUG
 {
     unsigned loopSize = 0;
 
+    JITDUMP("*************** In getLoopSize() for %s\n", emitLabelString(igLoopHeader), loopSize);
```
Comment (BruceForstall) [inline src/coreclr/jit/emit.cpp:5146]: `loopSize` doesn't have a printf format specifier.

File: src/coreclr/jit/emit.cpp [hunk 4]
```diff
@@ -5078,6 +5079,52 @@ bool emitter::emitEndsWithAlignInstr()
...
+    {
+        // If we never removed align flag of any loop or if
+        // the IG already has an align flag, it means it was not removed.
+        return false;
+    }
+
+    // Otherwise scan the align instruction list and see if the
+    // `igToCheck` was present.
+    instrDescAlign* alignInstr = emitAlignList;
+    while ((alignInstr != nullptr))
```
Comment (BruceForstall) [inline src/coreclr/jit/emit.cpp:5114]: Wouldn't this be simpler as a `for` loop?

File: src/coreclr/jit/emit.cpp [hunk 5]
```diff
@@ -5078,6 +5079,52 @@ bool emitter::emitEndsWithAlignInstr()
...
+//  emitIGHadAlignFlag: Checks if 'igToCheck' ever was marked as aligned and later
+//                      decided to not align. Sometimes, a loop is marked as not
+//                      needing alignment, but the igSize was not adjusted immediately.
+//                      This method would helpful to know that specially during loopSize
+//                      calculation where we would adjust the loopsize by removed alignment
+//                      bytes.
+//
+//  Returns:  true iff igToCheck had alignment but was later removed. For any other IG
+//            whether was marked aligned or not, the method will return false.
+//
```
Comment (BruceForstall) [inline src/coreclr/jit/emit.cpp:5101]: Couldn't you just have a persistent flag on an IG that marks it as having ever been marked as needing alignment, so you wouldn't need to search the list?

And if you remove alignment from a block, shouldn't you remove it from the list?

**Review by BruceForstall — CHANGES_REQUESTED** [review]

File: src/coreclr/jit/compiler.cpp (hunk 2 above)
Comment (kunalspathak) [reply src/coreclr/jit/compiler.cpp:5157]: Yes, probably I should use that.

File: src/coreclr/jit/emit.cpp (hunk 5 above)
Comment (kunalspathak) [reply src/coreclr/jit/emit.cpp:5101]: > And if you remove alignment from a block, shouldn't you remove it from the list?

We will set `IGF_REMOVED_ALIGN` only during `emitSetLoopBackEdge()`. At that time, we don't remove the `alignInstr` from the list nor do we adjust the `igSize` and `igOffs`. We do it later during `emitLoopAlignAdjustments()` phase. If we remove the `alignInstr` from the list, we will not process it and won't know that we need to readjust the offset and size of IGs that follows this one.

**Review by BruceForstall — CHANGES_REQUESTED** [review]

File: src/coreclr/jit/clrjit.natvis [hunk 6]
```diff
@@ -28,8 +28,8 @@ Documentation for VS debugger format specifiers: https://docs.microsoft.com/en-u
 
   <Type Name="Compiler::LoopDsc">
     <DisplayString Condition="lpFlags &amp; LPFLG_REMOVED">REMOVED</DisplayString>
-    <DisplayString Condition="lpFlags &amp; LPFLG_HAS_PREHEAD">BB{lpTop->bbNum,d}-BB{lpBottom->bbNum,d} pre-h:BB{lpHead->bbNum,d} e:BB{lpEntry->bbNum,d} {lpFlags,en}</DisplayString>
-    <DisplayString>BB{lpTop->bbNum,d}-BB{lpBottom->bbNum,d} h:BB{lpHead->bbNum,d} e:BB{lpEntry->bbNum,d} {lpFlags,en}</DisplayString>
+    <DisplayString Condition="lpFlags &amp; LPFLG_HAS_PREHEAD">BB{lpTop->bbNum,d}-BB{lpBottom->bbNum,d} pre-h:BB{lpHead->bbNum,d} e:BB{lpEntry->bbNum,d} [BB{lpTop->bbNum,d}..BB{lpBottom->bbNum,d}] {lpFlags,en}</DisplayString>
+    <DisplayString>BB{lpTop->bbNum,d}-BB{lpBottom->bbNum,d} h:BB{lpHead->bbNum,d} e:BB{lpEntry->bbNum,d} [BB{lpTop->bbNum,d}..BB{lpBottom->bbNum,d}] {lpFlags,en}</DisplayString>
```
[the remark this answers is not in the document]
Comment (kunalspathak) [reply src/coreclr/jit/clrjit.natvis:32]: I think I was trying to display something else and ended up with using `lpTop` and `lpBottom`. I will revert this and reformat the text a little bit.

**Review by BruceForstall — APPROVED** [review]

**Pull request merged**